File: current_session/5.py
"""
5. Longest Palindromic Substring
Medium

Given a string s, find the longest palindromic s in s.
You may assume that the maximum length of s is 1000.

Example 1:

Input: "babad"
Output: "bab"
Note: "aba" is also a valid answer.

Example 2:

Input: "cbbd"
Output: "bb"
"""
class Solution(object):    
    def longestPalindrome(self, s):
        """
        :type s: str
        :rtype: str
        """
        # dp[i][j] = True if palindrome
        dp = self.initialize(len(s))
        maxLen, start, end = 0, 0, 0
        for i in xrange(len(s), -1, -1):
            for j in xrange(i, len(s)):
                dp[i][j] = dp[i+1][j-1] and s[i] == s[j]
                if dp[i][j] and (j-i) > maxLen:
                    maxLen, start, end = j-i, i, j
        return s[start:end+1]

    def initialize(self, length):
        dp = [[-1] * (length+1) for _ in xrange(length+1)]
        for i in xrange(length-1):
            dp[i][i+1] = True
        return dp

print Solution().longestPalindrome("abcda"), "a"
print Solution().longestPalindrome("abccba"), "abccba"
print Solution().longestPalindrome("aaaaaaaaaaa"), "aaaaaaaaaaa"
print Solution().longestPalindrome("abba"), "abba"
print Solution().longestPalindrome("ac"), "a"
print Solution().longestPalindrome(""), ""

File: old_session/session_1/_5/c++/_5.cpp
/* Copyright [2017] <YJLee>

5. Longest Palindromic Substring

    Total Accepted: 169798
    Total Submissions: 686389
    Difficulty: Medium
    Contributors: Admin

Given a string s, find the longest palindromic substring in s.
You may assume that the maximum length of s is 1000.

Example:

Input: "babad"
Output: "bab"

Note: "aba" is also a valid answer.

Example:

Input: "cbbd"
Output: "bb"

*/
#include <iostream>
#include <string>

using std::string;
using std::cout;
using std::endl;

class Solution {
 public:
    std::string longestPalindrome(std::string s) {
        if (s.size() < 2)
            return s;
        int wordlen = s.size(), start = 0, pldlen = 1, left, right;
        for (int i = 0; (i + pldlen / 2) < wordlen;) {
            left = right = i;
            while (right < wordlen - 1 && s[right + 1] == s[right])
                ++right;
            i = right + 1;
            while (right < wordlen - 1 && left > 0 && s[right + 1] == s[left - 1]) {
                ++right;
                --left;
            }
            if (pldlen < right - left + 1) {
                start = left;
                pldlen = right - left + 1;
            }
        }
        return s.substr(start, pldlen);
    }
};

int main(int argc, const char * argv[]) {
    Solution sol = Solution();
    std::string word = "babad";
    std::string word2 = "a";
    std::string word3 = "uriwhfooxooxxookkkkkfjshpopopopopopopopopiiiiipopopopojjjj";
    std::cout << word << " => " << sol.longestPalindrome(word) << std::endl;
    std::cout << word2 << " => " << sol.longestPalindrome(word2) << std::endl;
    std::cout << word3 << " => " << sol.longestPalindrome(word3) << std::endl;
    return 0;
}
